tools/python/stylesheet/color_matrix_builder.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python2.7
+
+from __future__ import print_function
+
+import drules_struct_pb2 as proto
+
+import re
+import sys
+
+from collections import defaultdict
+from image_maker import argb_to_tuple, ImageMaker
+
+from os import path
+from protobuf_util import *
+
+PRIORITY_RE = re.compile("(.*)__(-?\d*)/(.*)")
+
+
+
+class MatrixBuilder:
+    def __init__(self):
+        self.transformation_table = defaultdict(list)  # {(color1, color2): [Element]}
+        self.inverted_transformation_table = {}  # {element_key: Number_in_transformation_table}
+        self.paths = []
+        self.drules = []
+        self.imgs = []
+        self.all_elements = defaultdict(lambda: defaultdict(lambda: -1))  # {path_index: {[(el.name, el.priority), (...)] : colour}}
+        self.key_correspondences = defaultdict(dict)  # {path_index: {old_element_path: new_element_path, ...}, ...}
+
+
+    # ADDING THE FILE:
+    def add_file(self, filepath):
+        print("Adding file: {}".format(filepath))
+        self.paths.append(filepath)
+        self.drules.append(self.read_bin(filepath))
+
+
+    def read_bin(self, filepath):
+        drules = proto.ContainerProto()
+        with open(filepath) as infile:
+            drules.ParseFromString(infile.read())
+        return drules
+
+
+    # READING:
+    def reader_factory(self, index):
+        def reader(sub_drules, index_list):
+            for key, _, color in self.color_key_generator(sub_drules, index_list):
+                self.all_elements[index][key] = color
+        return reader
+
+
+    def parse_all_bins(self):
+        print("Parsing the bin files")
+        self.foreach_path(lambda index: self.parse_bin(index))
+
+
+    def parse_bin(self, index):
+        self.recursive_iterator(self.drules[index], "", index, self.reader_factory(index))
+
+
+    # CALCULATE COLOR MATRIX
+    def calculate_color_matrix(self):
+        self.normalize_all_keys()
+        all_keys = set.union(*map(lambda s: set(s[1]), self.all_elements.iteritems()))
+        colors_table = defaultdict(list)  # {key: [color_0, color_1, -1 if element absent in style 2, ...]}
+
+        for key in all_keys:
+            for path_index in xrange(len(self.paths)):
```

Comment:
I don't get this one. 

colors_table is a dictionary of lists, why would we want to put a tuple in it?